docs/DEVELOPMENT-TIPS-TRICKS.md

```diff
@@ -0,0 +1,184 @@
+- Include assertions for matching snapshot
+
+```js
+test('Produces correct JSON output structure', () => {
+  expect(result).toMatchSnapshot();
+});
+```
+
+- Run [jest in watch mode][9] to iterate faster. Each time you edit the file with
+  the transform function, the test runner will restart the test. You might want to
+  use `console.log` to see results of your transform function while you develop it.
+
+### Debugging a lambda function locally
+
+As AWS lambda functions are also known as cloud functions, they are not always easy
+to debug locally. However, making use of the serverless framework, we have some helpers
+to do debugging locally.
+
+1.  In `webpack.config.js`, set a new property if not already set `devtool: 'source-map'`.
+    This will let you to see code for humans, not transpiled one during execution of code
+
+2.  Add a breakpoint in with `debug;` either in lambda function or transform function
+
+3.  Run the following in the CLI
+    `node --inspect-brk ./node_modules/.bin/serverless invoke local -f fooFunction --path eventStub.json`
+
+Where:
+
+- `node --inspect-brk` is the Node.js core inspector protocol which opens a session to Chrome browser
+- `serverless invoke local` is a specific command of the [serverless CLI][11]
+- `fooFunction` is the name of the function as found in the `serverless.yml` file
+- `eventStub` is the JSON file which you can take by using `console.log(JSON.strigify(event))`
+  in the beginning of the lambda function
+
+Open [Node.js debugger your Chrome browser's devtools][12]. You can also use the
+REPL with core debugger as well.
+
+As a result, you will be able step in and debug your cloud function locally with a decently
+close emulation to the real environment.
+
+![Serverless debugging](./assets/serverless-debugging.jpg)
+
+**Note**
+
+Because the lambda function of a given ETL depends on a file being present on S3 and
+the transform function is meant to input this file and output another file with normalized
+structure, do not remove the given S3 file during the debugging phase. This way, the
+resource you want to test will be an actual file existing in the cloud and the lambda
+function will operate quite closer to real life.
+
+Another close resemblance to S3 read stream of files is to use the Node.js's core
+[fs.createReadStream(path[, options])][13]. If you use this swapping approach you win
+independence of AWS services, but you lose possible system-specific behaviors of
+AWS SDK.
+
+## Use TDD for faster iterations on custom features
+
+Apart from using unit tests for faster development of ETLs, TDD can be also useful
+for the faster development of any other type of custom logic which does not rely on
+AWS functions.
+
+In fact, we do sometimes mock AWS services, but these tests have far less value in
+comparison to custom code which contain specific business logic and is more important
+to be maintainable. AWS services are not only maintained by AWS, but they could also
```

Comment:
> AWS services are not only maintained by AWS, but they could also change in time.

what do you mean?

Reply:
APIs can change and it doesn't depend on us, so no need to imitate code which we don't own.